### PR TITLE
Update rules_haskell

### PR DESCRIPTION
**Pull request by aherrmann-da**
- rules_haskell now handles the global package db within Bazel tweag/rules_haskell#859
- We no longer use the Nix provided c2hs. So, we drop it.
- Rename `ghcWithC2hs` to `ghcStatic` to clarify that that's where the static linking patches are applied.
- Extend package-db patches to align Nix store paths with the new `$out`. This works around a restriction in current rules_haskell, where the paths in the package config files must have the same prefix as the path to the package config files themselves. (This is also related to why we need to pull in libffi headers separately on Unix)
- Adapt static linking patches to new rules_haskell
- Simplify libffi headers on Windows. On Windows they are part of the GHC bindist and are now picked up by the new global package db handling, so we don't need to provide an additional target anymore.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: bazel_tools/haskell-keep-hs-extra-libs.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+have to undo those patches.
+diff --git a/haskell/private/actions/link.bzl b/haskell/private/actions/link.bzl
+index cb6aa19..65cd2c6 100644
+--- a/haskell/private/actions/link.bzl
++++ b/haskell/private/actions/link.bzl
+@@ -485,15 +485,6 @@ def _add_external_libraries(args, ext_libs):
+     deduped = list.dedup_on(get_lib_name, ext_libs)
+ 
+     for lib in deduped:
+-        # This test is a hack. When a CC library has a Haskell library
```
Comment (cocreature) [inline bazel_tools/haskell-keep-hs-extra-libs.patch]: Not sure if the fact that we now have patches to remove hacks in rules_haskell is a good or a bad thing :smile:

File: bazel_tools/haskell-package-env.patch [hunk 2]
```diff
@@ -1,5 +1,5 @@
 diff --git a/haskell/private/actions/compile.bzl b/haskell/private/actions/compile.bzl
```
Comment (cocreature) [inline bazel_tools/haskell-package-env.patch]: We should try to upstream that patch. Just haven’t had the time so far sadly :disappointed:

File: nix/ghc.nix [hunk 3]
```diff
@@ -24,7 +24,6 @@ let
 
       # We need newer versions that build with GHC 8.6
       language-c = super.callPackage ./overrides/language-c-0.8.2.nix {};
```
Comment (cocreature) [inline nix/ghc.nix]: We can probably also get rid of the `language-c` override if we don’t need `c2hs`.

Comment (neil-da) [reply nix/ghc.nix]: I believe we need it for grpc related stuff.

Comment (aherrmann-da) [reply nix/ghc.nix]: The Nix override seems to be unused. I could only find references to the Hazel package.

Comment (cocreature) [reply nix/ghc.nix]: :+1:

File: bazel_tools/haskell-package-env.patch (hunk 2 above)
Comment (aherrmann-da) [reply bazel_tools/haskell-package-env.patch]: I tried to just apply the patch on rules_haskell. Sadly it breaks plugins, haddock, and doctests.

**Pull request merged**